### PR TITLE
Listen for debugger shell in container

### PR DESCRIPTION
**Pull request by alexcb**
Adds a shell repeater server which listens for tcp connections from both
the debugger and earth command.

The shell repeater had to be introduced because sockets do not work on
macOS with docker; and exposing a listening port did not work when a
firewall was in place.

### CODE REVIEW

**Comment by alexcb** [comment]
Here's an example of it working on macOS

image

File: buildkitd/entrypoint.sh [hunk 1]
```diff
@@ -114,4 +114,25 @@ if [ "$use_loop_device" == "true" ]; then
...
 fi
 
-exec "$@"
+# start shell repeater server
+echo starting shellrepeater
+which shellrepeater
+shellrepeater &
+shellrepeaterpid=$!
+
+exec "$@" &
```
[the remark this answers is not in the document]
Comment (alexcb) [reply buildkitd/entrypoint.sh:123]: I tried it and it worked like this, as does:
```
#!/bin/sh

exec ls &
echo hello
```

however, it's really not needed, so i'll remove it :smile:

File: debugger/server/server.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+	dataForShell    chan []byte
+	dataForTerminal chan []byte
+
+	sigs chan os.Signal
+	addr string
+}
+
+func (s *Server) handleConn(conn net.Conn, readFrom, writeTo chan []byte) {
+	ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (alexcb) [reply debugger/server/server.go:28]: good catch.

File: buildkitd/entrypoint.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (vladaionescu) [reply buildkitd/entrypoint.sh]: Ah ok, interesting!

**Pull request merged**